### PR TITLE
[SYCL][Driver] Add `off` value for device code split option

### PR DESCRIPTION
**Pull request by Fznamznon**
Signed-off-by: Mariya Podchishchaeva <mariya.podchishchaeva@intel.com>

### CODE REVIEW

**Comment by Fznamznon** [comment]
Sorry, I forgot the test for https://github.com/intel/llvm/pull/853 . My bad :(

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -3575,7 +3575,7 @@ class OffloadingActionBuilder final {
       Arg *SYCLLinkTargets = Args.getLastArg(
                                   options::OPT_fsycl_link_targets_EQ);
       WrapDeviceOnlyBinary = Args.hasArg(options::OPT_fsycl_link_EQ);
-      DeviceCodeSplit = Args.hasArg(options::OPT_fsycl_device_code_split);
+      DeviceCodeSplit = Args.hasArg(options::OPT_fsycl_device_code_split_EQ);
```
Comment (bader) [inline clang/lib/Driver/Driver.cpp]:
```suggestion
      DeviceCodeSplit = Args.getLastArg(options::OPT_fsycl_device_code_split_EQ);
```

Comment (Fznamznon) [reply clang/lib/Driver/Driver.cpp]: Done.

**Comment by Fznamznon** [comment]
Does anybody know why all the sycl driver tests are xfailed on windows? Even windows-specific test. https://github.com/intel/llvm/blob/a339d4cf48200ea879a8740478201cfbeb704fca/clang/test/Driver/sycl-offload-win.c#L101
The test doesn’t seem failing with split-specific problem.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp]: `getLastArg()` yeilds `Arg *`, which is then implicitly converted to `bool`. If the aim is to check for `-fsycl-device-code-split=` value, `DeviceCodeSplit`'s type had better be turned into a string representation, which can be retrieved by `Arg::getValue()`, or set to `""` if `getLastArg()` yields `nullptr`.

**Comment by AGindinson** [comment]
> Does anybody know why all the sycl driver tests are xfailed on windows? Even windows-specific test.
> 
> https://github.com/intel/llvm/blob/a339d4cf48200ea879a8740478201cfbeb704fca/clang/test/Driver/sycl-offload-win.c#L101
> 
> 
> My test doesn’t seem failing with split-specific problem.

I can't analyze this properly now, but it seems related to some Windows-unfriendly path regexp.

Definitely not caused by your patch, but I'd like some input from @mdtoguchi or @v-klochkov before we blissfully add the same `TODO + XFAIL` and commit the test.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Driver/Driver.cpp]: `DeviceCodeSplit` variable is needed here only to check presence of device code split option. Value is checked in SYCL toolchain. I think I can try implement request from @bader and add `none` value for `-fsycl-device-code-split` option, then conversion `DeviceCodeSplit`'s type to a string representation will make sense.

**Comment by Fznamznon** [comment]
> Definitely not caused by your patch, but I'd like some input from @mdtoguchi or @v-klochkov before we blissfully add the same `TODO + XFAIL` and commit the test.

I wanted to check that all tools is called properly when device code split is requested. So I took existing test cases and extended these test cases for device code split case. The problem happens in lines not related to split, so another option is reduce test cases to make them check only split-affected jobs.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp]: Looks reasonable. There's also an option to leave `DeviceCodeSplit` boolean and set it like this:
```
auto *DeviceCodeSplitArg = Args.getLastArg(OPT_fsycl_device_code_split_EQ);
DeviceCodeSplit = DeviceCodeSplitArg && DeviceCodeSplitArg->getValue() != "none";
```

**Comment by AGindinson** [comment]
> The problem happens in lines not related to split, so another option is reduce test cases to make them check only split-affected jobs.

Given the feature complexity, I'd personally steer away from reducing coverage. The least risky option seems to be `XFAIL: windows-msvc`, and a follow-up update with either making the test "split-only" or  fixing the Windows problem, depending on others' opinions.

**Comment by Fznamznon** [comment]
> > The problem happens in lines not related to split, so another option is reduce test cases to make them check only split-affected jobs.
> 
> Given the feature complexity, I'd personally steer away from reducing coverage. The least risky option seems to be `XFAIL: windows-msvc`, and a later follow-up with either making the test "split-only" or analyzing + fixing the Windows problem, depending on others' opinions.

How about add XFAIL and create a GitHub issue?

**Comment by AGindinson** [comment]
> How about add XFAIL and create a GitHub issue?

That's a good idea, I'll create one.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Driver/Driver.cpp]: Done.

**Comment by Fznamznon** [comment]
> > How about add XFAIL and create a GitHub issue?
> 
> That's a good idea, I'll create one.

Okay then, I XFAILed the test.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp]: I understand that no additional handling of `none` is needed at the `SYCLToolchain` level, but could you please confirm this?

Comment (Fznamznon) [reply clang/lib/Driver/Driver.cpp]: By what?

Comment (AGindinson) [reply clang/lib/Driver/Driver.cpp]: By a comment akin to "Yes, I've checked myself and `none` doesn't need to be handled there", or by a new commit that does take it into account... Apparently :)

File: clang/include/clang/Driver/Options.td [hunk 2]
```diff
@@ -1821,7 +1821,7 @@ def fsycl_link_targets_EQ : CommaJoined<["-"], "fsycl-link-targets=">, Flags<[Dr
   HelpText<"Specify comma-separated list of triples SYCL offloading targets to produce linked device images">;
 def fsycl_device_code_split_EQ : Joined<["-"], "fsycl-device-code-split=">,
    Flags<[CC1Option, CoreOption]>, HelpText<"Perform SYCL device code split: per_kernel (device code module is "
-  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)). "
+  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | none (no device code split). "
```
Comment (bader) [inline clang/include/clang/Driver/Options.td]:
```suggestion
  "created for each SYCL kernel) | per_source (device code module is created for each source (translation unit)) | off (no device code split). "
```

File: clang/include/clang/Driver/Options.td [hunk 3]
```diff
@@ -1821,7 +1821,7 @@ def fsycl_link_targets_EQ : CommaJoined<["-"], "fsycl-link-targets=">, Flags<[Dr
...
   "Default is no device code split - all kernels go into a single module`">, Values<"per_source,per_kernel">;
```
Comment (bader) [inline clang/include/clang/Driver/Options.td]:
```suggestion
  "Default is 'off' - all kernels go into a single module`">, Values<"per_source, per_kernel, off">;
```

File: clang/include/clang/Driver/Options.td (hunk 2 above)
Comment (Fznamznon) [reply clang/include/clang/Driver/Options.td]: Done.

File: clang/include/clang/Driver/Options.td (hunk 3 above)
Comment (Fznamznon) [reply clang/include/clang/Driver/Options.td]: Done.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Driver/Driver.cpp]: Yes, I've checked myself :)

**Pull request merged**